### PR TITLE
[DEPRECATED] Proto changes to support explicit disclosure in conjunction with template subscriptions

### PR DESCRIPTION
**Pull request by mziolekda**
Two changes to the proto definitions

- Make the blob mandatory on the command submission
- Extend template filter to allow for requesting blobs when subscribing for templates

Why is this change necessary:

- Either a publisher or a client in a typical explicit-disclosure scenario may be based on an old package version that is missing a couple of fields. As a result some of the create-arguments may be dropped in the communication. This would render the subsequent command submission a failure. This is why a blob MUST be a required field.
- The typed arguments on the other hand CAN be an optional filed. A client needs to trust the publisher of the data that the contract it has received off ledger is legit. Alas, there is no way of validating of what is in the blob and that the blob and the typed arguments match. The only one that can verify that is ledger api server. It can do it when validating the submission message.

### CODE REVIEW

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto [hunk 1]
```diff
@@ -219,16 +219,15 @@ message DisclosedContract {
   // Required
   string contract_id = 2;
 
-  // The contract arguments
+  // The contract arguments as typed Record
   // Required
```
Comment (tudor-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:223]: This one can be optional no?

Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:223]: Also we should probably not reuse the proto indexes

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto [hunk 2]
```diff
@@ -56,6 +56,16 @@ message InclusiveFilters {
...
+
+  // A collection of templates for which the payload will be included in the
+  // ``create_arguments`` of a matching ``CreatedEvent``.
+  // SHOULD NOT contain duplicates.
+  // In contrast to the ``template_ids`` field, one can specify whether the blob should be provided
+  // alongside the create arguments.
+  // All ``template_id`` inside need to be valid: corresponding template should be defined in one of
+  // the available packages at the time of the query.
+  // Optional
+  repeated TemplateFilter template_flters = 3;
```
Comment (andreaslochbihler-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:68]: Do you want to specify the resolution rules if the interface_filters contradict the template_filters, e.g., an interface_filter specifies to not `include_create_arguments_blob` and the `template_filters` do so for some templates that implement this interface? IMO we should OR those conditions.

Comment (andreaslochbihler-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:68]:
```suggestion
  // A collection of templates for which the payload will be included in the
  // ``create_arguments`` of a matching ``CreatedEvent``.
  // SHOULD NOT contain duplicates.
  // In contrast to the ``template_ids`` field, one can specify whether the blob should be provided
  // alongside the create arguments.
  // All ``template_id`` inside need to be valid: corresponding template should be defined in one of
  // the available packages at the time of the query.
  // Optional
  repeated TemplateFilter template_filters = 3;
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 1 above)
Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:223]: > Also we should probably not reuse the proto indexes

As discussed offline, probably not an issue if it doesn't break wire backwards compatibility

Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:223]: Yes, apparently this works nicely in proto

- new server can deal with old client which will only populate one of the fields
- old server will pick the later field (blob=5) when confronted with both

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto (hunk 2 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:68]: In fact this problem already exists for the interfaces alone, because a template can implement two of them. It is good however to spell out the rules here. 
:+1:

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 1 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:223]: I have changed this now to be backwards compatible Apparently we have enough downstream components to make this a necessity

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto (hunk 2 above)
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto]:
```suggestion
  repeated TemplateFilter template_filters = 3;
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto [hunk 3]
```diff
@@ -77,3 +87,18 @@ message InterfaceFilter {
   // Optional
   bool include_create_arguments_blob = 3;
 }
+
+// This filter matches contracts that implement a specific interface.
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:94]:
```suggestion
// This filter matches contracts of a specific template.
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto [hunk 4]
```diff
@@ -219,16 +219,15 @@ message DisclosedContract {
...
+  Record create_arguments = 3;
 
-    // The contract arguments specified using an opaque blob extracted from the ``create_arguments_blob`` field
-    // of a ``com.daml.ledger.api.v1.CreatedEvent``.
-    google.protobuf.Any create_arguments_blob = 5;
-  }
+  // The contract arguments specified using an opaque blob extracted from the ``create_arguments_blob`` field
+  // of a ``com.daml.ledger.api.v1.CreatedEvent``. Ledger API server validates this data in conjunction with
+  // the ``create_arguments`` record and returns a comprehensive error when any inconsistency is found.
+  // Required
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: Why is this required?

Can't I use either of the two, and if I use both get additional validation applied?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto [hunk 5]
```diff
@@ -56,6 +56,16 @@ message InclusiveFilters {
   // the available packages at the time of the query.
   // Optional
   repeated InterfaceFilter interface_filters = 2;
+
+  // A collection of templates for which the payload will be included in the
+  // ``create_arguments`` of a matching ``CreatedEvent``.
+  // SHOULD NOT contain duplicates.
+  // In contrast to the ``template_ids`` field, one can specify whether the blob should be provided
+  // alongside the create arguments.
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:64]: What happens if a `template_ids` entry is equal to one of the template filters?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 4 above)
Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: Why do we even need this change away from the `oneof`?

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: > Make the blob mandatory on the command submission

This is expensive for consumers like scribe (aka PQS): they now have to always store both the expanded form of the create arguments (for indexing) and the full blob for all templates they subscribe to. Thereby requiring double the amount of storage.

What's the rationale for this change?

**Review by meiersi-da — CHANGES_REQUESTED** [review]
@mziolekda I'm probably missing context, but don't yet understand the rationale for this change. Could you explain why we need this change?

FYI: @moritzkiefer-da

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto (hunk 2 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto]: I had that corrected in a subsequent commit, but couldn't merge due to borked repo

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 4 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: Only the blob will be mandatory and is a consequence of the upgrading story. The client will likely use a code-generated stub to handle the contracts which may be based on an older version that will drop some of the contract fields

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: OK. I can see that upgrading might require always passing `create_argument_blob`s for disclosed contracts.

Do you have a design doc clarifying the exact requirement(s) and why this solution solves it, as well as the limitations (like the 2x storage increase for PQS for contracts that need to be disclosed?).

Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: I have subsequently changed the PR to allow for backwards compatibility. Again my change was stuck due to repo being borked

Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: We have an extension to the original ED document:
https://docs.google.com/document/d/1uhDEAKLOqVHXiNOUNcNtABlAzvo1ELRdnDEXOWVIQyg/edit#bookmark=id.2pwvqe594wr8

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto [hunk 6]
```diff
@@ -56,6 +56,18 @@ message InclusiveFilters {
...
+  // If a contract is simultaneously selected by a template filter and one or more interface filters,
+  // the corresponding ``include_create_arguments_blob`` are consolidated using an OR operation.
+  // All ``template_id`` inside need to be valid: corresponding template should be defined in one of
+  // the available packages at the time of the query.
+  // Optional
+  repeated TemplateFilter template_filters = 3;
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:70]: @mziolekda : wdyt of making the `template_ids` field deprecated?

cf. https://stackoverflow.com/questions/60490487/whats-the-best-way-to-deprecate-a-field-in-protocol-buffer-v3-reserved-vs-depre

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 4 above)
Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: Thanks for the clarifications @mziolekda . It looks like we're converging here. I've added a comment on the design doc.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto [hunk 7]
```diff
@@ -251,4 +251,16 @@ message DisclosedContract {
   // Maintainers of the contract key, if defined.
   // Optional
   repeated string key_maintainers = 9;
+
+  // The contract arguments as a typed record. The fields comprising the create arguments are cross-validated
+  // with the contents of the ``contract_create_arguments_blob``. If any inconsistency is determined,
+  // the command submission is rejected. It can be left unset.
+  // Optional
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:258]: @mziolekda : note that the same kind of checking is required for the `signatories`, `observers`, `contract_key`, and `key_maintainers`. Will that be done too? Consider moving the comment about the checks to the `contract_create_arguments_blob`.

Also consider going the alternative way suggested in https://docs.google.com/document/d/1uhDEAKLOqVHXiNOUNcNtABlAzvo1ELRdnDEXOWVIQyg/edit?disco=AAAA-QEyClA where only the minimal amount of data is passed in for a disclosed contract, and decoding is done via an explicit rpc offered by the Ledger API server.

This minimizes encoding problems. One that I can see lurking is actually the `contract_key` of type `Value`, which is likely as hard to pass via clients w/o loss of information as the `create_arguments`.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto (hunk 6 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:70]: Good point. For the templates we should definitely follow the `deprecated_template_ids` approach variant. I would like to do the same to the

```
oneof arguments {
    // The contract arguments as typed Record
    Record deprecated_create_arguments = 3;
    // The contract arguments specified using an opaque blob extracted from the ``create_arguments_blob`` field
    // of a ``com.daml.ledger.api.v1.CreatedEvent``.
    google.protobuf.Any deprecated_create_arguments_blob = 5;
  }
``` 
WDYT?

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:70]: I would not change the names. I'd just add `[deprecated = true]` and let the downstream build tools (e.g., `javac` and `scalac`) handle the generation of the deprecation message.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 7 above)
Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto:258]: Actually, this approach would also have the benefit of not requiring any change to the API. Just change the commentary that only the `create_arguments_blob` will be supported with upgrading, and mark the `create_arguments` with `[deprecated = true]`.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto (hunk 6 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:70]: the reason why I like changing the names, because right now there was a clash between `arguments.create_arguments` and `create_arguments`, which has forced me to rename the new ones to `contract_create_arguments`. Not nice :-( I much rather have nice names for new fields and crummy names for old ones.

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto:70]: See my comment here: https://github.com/digital-asset/daml/pull/17546#discussion_r1360396293

I believe we should give that a serious shot, and thereby also remove any need for introducing new fields.

**Comment by simonmaxen-da** [comment]
The ContractInstance is already used in the creation if the contract id suffix. This is checked on submission. For this reason any change to the contract argument as a result of marshalling/un-marshalling as code-generated bindings will result in an invalid contract.

To avoid this problem the client needs to always submit the presented, or stored, contract argument when submitting the command. By changing the bindings/educating developers to use disclosed contract arguments in this was will solve this issue iiuc.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto (hunk 4 above)
Comment (raphael-speyer-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto]: I don't think our code-gen types should drop unknown fields. Can we preserve them (without interpretation), and include them in any output format, similar to the protobuf behaviour?
